charybdis/disasm.py

```diff
@@ -20,51 +20,63 @@ def decode_insn(rom: bytes, index: int) -> Optional[DecodedInsn]:
             decoded = insn.Insn(name=insn.InsnName.NOP)
         # LD (BC), A
         case 0x02:
-            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.Indirect(insn.R16.BC), insn.R8.A])
+            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.IndirectR16(insn.R16.BC), insn.R8.A])
         # LD B, n
         case 0x06:
             size = 2
             imm = rom[index + 1]
             decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.B, insn.U8(imm)])
         # LD A, (BC)
         case 0x0A:
-            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.A, insn.Indirect(insn.R16.BC)])
+            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.A, insn.IndirectR16(insn.R16.BC)])
         # LD C, n
         case 0x0e:
             size = 2
             imm = rom[index + 1]
             decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.C, insn.U8(imm)])
         # LD (DE), A
         case 0x12:
-            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.Indirect(insn.R16.DE), insn.R8.A])
+            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.IndirectR16(insn.R16.DE), insn.R8.A])
         # LD D, n
         case 0x16:
             size = 2
             imm = rom[index + 1]
             decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.D, insn.U8(imm)])
         # LD A, (DE)
         case 0x1A:
-            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.A, insn.Indirect(insn.R16.DE)])
+            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.A, insn.IndirectR16(insn.R16.DE)])
         # LD E, n
         case 0x1e:
             size = 2
             imm = rom[index + 1]
             decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.R8.E, insn.U8(imm)])
+        # LD [HL+], A
+        case 0x22: 
+            decoded = insn.Insn(name=insn.InsnName.LD, operands=[insn.IndirectR16Index(insn.R16.HL, True), insn.R8.A])
```

Comment:
Can you create two types `IndirectHLInc` and `IndirectHLDec` instead? It feels weird attaching extra data to `IndirectR16` when most inhabitants won't use that value. It's better to have a proliferation of small, specific types here, because we will be using [this](https://pypi.org/project/multimethod/) eventually.